internal/service/pokemon.go

```diff
@@ -0,0 +1,143 @@
+package service
+
+import (
+	"errors"
+	"github.com/martin-bernal/go-bootcamp-project/internal/helper"
+	"github.com/martin-bernal/go-bootcamp-project/internal/interfaces"
+	"strconv"
+	"sync"
+
+	"github.com/martin-bernal/go-bootcamp-project/internal/entity"
+	"github.com/martin-bernal/go-bootcamp-project/internal/repository"
+)
+
+//LAYER DEFINITION
+type PokemonService interface {
+	// CreatePokemon Function to create new pokemon
+	CreatePokemon(pokeName string) error
+	// FindPokemon Function to find pokemon
+	FindPokemon(pokeId string) (entity.Pokemon, error)
+	// GetPokemons Function to get all pokemons
+	GetPokemons() ([]entity.Pokemon, error)
+	//	GetPokemonsAsync Function to get all pokemons using worker pool
+	GetPokemonsAsync(typeParam string, items int, itemsPerWorkers int) ([]*entity.Pokemon, error)
+}
+
+type pokemonService struct {
+	repo          repository.PokemonRepo
+	pokemonClient interfaces.Client
+}
+
+type pokeTask struct {
+	Pokemon *entity.Pokemon
+	TypeID  string
+	TaskRun func(pokemon *entity.Pokemon, typeId string)
+}
+
+func (t pokeTask) Run() {
+	t.TaskRun(t.Pokemon, t.TypeID)
+}
+
+func NewPokemonService(repo repository.PokemonRepo, pokeClient interfaces.Client) PokemonService {
+	return &pokemonService{repo: repo, pokemonClient: pokeClient}
+}
+
+//IMPLEMENTATIONS
+
+func (ps *pokemonService) CreatePokemon(pokeName string) error {
+
+	pokemon, _ := ps.pokemonClient.GetPokemonFromApi(pokeName)
+
+	err := ps.repo.WritePokemon(pokemon)
+
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (ps *pokemonService) FindPokemon(pokeId string) (entity.Pokemon, error) {
+	pokemons, err := ps.repo.ReadPokemons()
+	var pokemon entity.Pokemon
+
+	pokeIntId, err2 := strconv.ParseInt(pokeId, 10, 64)
+	pokeFound := false
+
+	if err2 != nil {
+		return pokemon, errors.New("No valid int")
+	}
+
+	if err != nil {
+		return entity.Pokemon{}, err
+	}
+
+	for _, poke := range pokemons {
+		if poke.ID == pokeIntId {
+			pokemon = poke
+			pokeFound = true
+		}
+	}
+
+	if !pokeFound {
+		return pokemon, errors.New("Pokemon not found")
+	}
+
+	return pokemon, nil
+}
+
+func (ps *pokemonService) GetPokemons() ([]entity.Pokemon, error) {
+	pokemons, err := ps.repo.ReadPokemons()
+	if err != nil {
+		return nil, err
+	}
+	return pokemons, nil
+}
+
+func (ps *pokemonService) GetPokemonsAsync(typeParam string, items int, itemsPerWorkers int) ([]*entity.Pokemon, error) {
+	pokemons, err := ps.repo.ReadPokemons()
+	if err != nil {
+		return nil, err
+	}
+
+	pool := helper.NewGoroutinePool(3, itemsPerWorkers)
+
+	var res []*entity.Pokemon
```

Comment:
This is a terribe idea, since you are going to pass them to go routines, you should either use mutex to prevent two workers do something over the same index or get a race condition, or to use a channel to pass the information between go routines